src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs

```diff
@@ -0,0 +1,327 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Apple;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class AppleCrypto
+    {
+        [DllImport(Libraries.AppleCryptoNative)]
+        private static extern int AppleCryptoNative_SecKeychainItemCopyKeychain(
+            IntPtr item,
+            out SafeKeychainHandle keychain);
+        
+        [DllImport(Libraries.AppleCryptoNative, EntryPoint = "AppleCryptoNative_SecKeychainCreate")]
+        private static extern int AppleCryptoNative_SecKeychainCreateTemporary(
+            string path,
+            int utf8PassphraseLength,
+            byte[] utf8Passphrase,
+            out SafeTemporaryKeychainHandle keychain);
+
+        [DllImport(Libraries.AppleCryptoNative)]
+        private static extern int AppleCryptoNative_SecKeychainDelete(IntPtr keychain);
+
+        [DllImport(Libraries.AppleCryptoNative)]
+        private static extern int AppleCryptoNative_SecKeychainCopyDefault(out SafeKeychainHandle keychain);
+
+        [DllImport(Libraries.AppleCryptoNative)]
+        private static extern int AppleCryptoNative_SecKeychainOpen(
+            string keychainPath,
+            out SafeKeychainHandle keychain);
+
+        [DllImport(Libraries.AppleCryptoNative)]
+        private static extern int AppleCryptoNative_SecKeychainEnumerateCerts(
+            SafeKeychainHandle keychain,
+            out SafeCFArrayHandle matches,
+            out int pOSStatus);
+
+        [DllImport(Libraries.AppleCryptoNative)]
+        private static extern int AppleCryptoNative_SecKeychainEnumerateIdentities(
+            SafeKeychainHandle keychain,
+            out SafeCFArrayHandle matches,
+            out int pOSStatus);
+
+        internal static SafeKeychainHandle SecKeychainItemCopyKeychain(SafeKeychainItemHandle item)
+        {
+            var handle = SecKeychainItemCopyKeychain(item.DangerousGetHandle());
+            GC.KeepAlive(item);
+            return handle;
+        }
+
+        internal static SafeKeychainHandle SecKeychainItemCopyKeychain(IntPtr item)
+        {
+            SafeKeychainHandle keychain;
+            int osStatus = AppleCryptoNative_SecKeychainItemCopyKeychain(item, out keychain);
+
+            // A whole lot of NULL is expected from this.
+            // Any key or cert which isn't keychain-backed, and this is the primary way we'd find that out.
+            if (keychain.IsInvalid)
```

Comment:
The Marshaller guarantees that it never returns a (C#) null SafeHandle.  If it returns nullptr/NULL then it returns an object whose handle value is 0.  In this case, it will report as IsInvalid.

So this is a check for "did the native method return nullptr".